File: adt2unit/calc_pattern_pos.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

uint8_t calc_pattern_pos (uint8_t pattern)
{
  uint8_t r = BYTE_NULL; // result (pattern position)
  uint8_t n = 0; // jumps count
  int16_t i; // index

  DBG_ENTER ("calc_pattern_pos");

  i = calc_following_order (0);
  while ((i != -1) && (n + 1 < 0x80))
  {
    if (songdata.pattern_order[i] != pattern)
    {
      if (i + 1 < 0x80)
      {
        i = calc_following_order (i + 1);
        n++;
      } else
        break;
    }
    else
    {
      r = i;
      break;
    }
  }

  DBG_LEAVE (); //EXIT //calc_pattern_pos
  return r;
}

File: adt2unit/pas/calc_pattern_pos.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function calc_pattern_pos(pattern: Byte): Byte;

var
  index: Integer;
  jump_count,pattern_pos: Byte;

begin
  _dbg_enter ({$I %FILE%}, 'calc_pattern_pos');

  pattern_pos := BYTE_NULL;
  jump_count := 0;
  index := calc_following_order(0);
  While (index <> -1) and (jump_count < $7f) do
    If (songdata.pattern_order[index] <> pattern) then
      If NOT (index < $7f) then BREAK
      else begin
             Inc(index);
             index := calc_following_order(index);
             Inc(jump_count);
           end
    else begin
           pattern_pos := index;
           BREAK;
         end;

  calc_pattern_pos := pattern_pos;

  _dbg_leave; //EXIT //calc_pattern_pos
end;
